[cmake] ADDONS_TO_BUILD exclusion regex and better exact match handling

fuzzard opened the pull request:
## Description

**Commit 1** looks to resolve https://github.com/xbmc/xbmc/issues/17773

examples

```
ADDONS_TO_BUILD='game.libretro peripheral.*'"
```
This will match game.libretro (and only game.libretro), but also match peripheral.joystick and peripheral.xarcade

```
ADDONS_TO_BUILD='game.libretro.* game.libretro peripheral.*'"
```
This will match game.libretro exactly, everything in game.libretro (regex), and both peripheral addons

```
ADDONS_TO_BUILD='game.libretro.mame.* game.libretro peripheral.*'"
```
this will match game.libretro.mame, mame2000, mame2003 (7 mame addons in kodi-game binary addon repo), etc, game.libretro (exact) and then both peripheral addons

```
ADDONS_TO_BUILD='game.libretro.mame game.libretro peripheral.*'"
```
this will match game.libretro.mame (exact), game.libretro (exact) and then both peripheral addons


implements an exclusion match regex to exclude an addon/s from building

example usage
```
ADDONS_TO_BUILD='game.libretro.* -game.libretro.mrboom peripheral.joystick'
```

cmake output
```
-- Exclusion found -game.libretro.2048
-- Addon game.libretro.2048 matches exclusion rule -game.libretro.2048
-- Pattern game.libretro.* matches game.libretro.mrboom, building addon
```
game.libretro.mrboom will not build, but game.libretro and game.libretro.2048 will

example 2:
```
ADDONS_TO_BUILD='game.libretro -game.libretro.mame peripheral.joystick'
```
Cmake output
```
Showing Recent Messages
-- game.libretro.lutro DEPENDENCIES: lutro
-- Addon game.libretro.mame matches exclusion rule game.libretro.mame
-- Addon game.libretro.mame2000 matches exclusion rule game.libretro.mame
-- Addon game.libretro.mame2003 matches exclusion rule game.libretro.mame
-- Addon game.libretro.mame2003_plus matches exclusion rule game.libretro.mame
-- Addon game.libretro.mame2010 matches exclusion rule game.libretro.mame
-- Addon game.libretro.mame2015 matches exclusion rule game.libretro.mame
-- Addon game.libretro.mame2016 matches exclusion rule game.libretro.mame
-- Pattern game.libretro matches game.libretro.melonds, building addon
-- 
```

**Commit 2** Updates docs

## Motivation and Context
Some of the libretro cores can have long build times (eg. game.libretro.mame*).
This allows the ability to build selectively using regex (game.libretro.*) whilst excluding particular addon groups inside that initial regex

## How Has This Been Tested?
ios building kodi-game binary-addon repo and master binary-addon repo

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

AlwinEsch reviewed (changes requested):
About first way on here
-----

Have tested and works nice :smile:

Was on begin not sure if `if(exclusion MATCHES "^[-]${id}")` handle it correct, in case 2 addons have mostly same name e.g. `visualization.milkdrop` and `visualization.milkdrop2`, but has worked on test.

Documentation
-----
The only reason why I marked with "Request changes" is to add somewhere a documentation about, so that a user know this support.
Thought for here:
- https://github.com/xbmc/xbmc/blob/master/cmake/addons/README.md
- And maybe on them here https://github.com/xbmc/xbmc/tree/master/docs
  - not sure how, maybe by reference to `cmake/addons/README.md`

e.g. this on build docs:
```
(3) Build a specific group of add-ons:

make -j$(getconf _NPROCESSORS_ONLN) -C tools/depends/target/binary-addons ADDONS="pvr.*"

For set available styles of ADDONS you can also look by ADDONS_TO_BUILD on here [Kodi add-ons CMake based buildsystem](../cmake/addons/README.md)
```
Or take direct the text from the `cmake/addons/README.md` and copy to this files.

fuzzard commented on cmake/addons/CMakeLists.txt:
```diff
@@ -244,15 +244,40 @@ endif()
+  string(REGEX MATCH "^[-].*" FOUND_EXCLUSION "${addon}")
+  if(NOT FOUND_EXCLUSION STREQUAL "")
+    # Remove beginning - and then append to excluded addon list
+    STRING(REGEX MATCH "-([^ ]*)"  excluded_addon ${FOUND_EXCLUSION})
```
yeah, sorry, had a force push to fix it. Came across the exact issue you first mentioned.

Now i save the excluded_addon list as just eg pvr.dvblink;pvr.dvbviewer without the leading -. Makes the match in the actual addon loop simpler.

fuzzard replied:
the match is just the way i see regex matching. Its very loose, and doesnt require regex terms to explicitly force a regex search. ie. -pvr.dvb will match anything with pvr.dvb. Same goes for actual matching ie games.libretro actually matches games.libretro and all games.libretro.*

Honestly i havent tried to fix that situation.

AlwinEsch commented:
Yes e.g. about example 2, it would be better to have them as `-game.libretro.mame*` and to set with `*` that text is behind.
Without it can have problems that a for creation wanted addon can also be excluded.

fuzzard committed:
[docs] Provide further information regarding addon_to_build regex usage

fuzzard commented:
The current matches behaviour hasnt changed with this PR.
At this time, i would see that as out of scope of this PR. I definitely know what you are talking about, but it is also existing behaviour due to the way we use MATCHES

https://github.com/xbmc/xbmc/blob/35641bab4e2d97340269795d0a3a5144aa7538e1/cmake/addons/CMakeLists.txt#L261

The way i see it, you would need to check the supplied string ends in a regex search term (ie. .*), otherwise only do the strict STREQUALS check. Sure it can be done, but it will change current behaviour. Not sure i personally want to get involved in bulk changes, haha.

phunkyfish commented:
I think just be sure the docs reflect the actual behaviour and that's ok.

garbear reviewed (approved):
Might not have a chance to review the CMake this week but +1 for general concept

fuzzard commented:
@AlwinEsch when you have time could you have a look at https://github.com/fuzzard/xbmc/commit/f479fb7a0fe6ed57c7535f2fc831ebeb8dae2646

Trying to come up with a way to no regex match a generic term (game.libretro, and vis.milkdrop i your scenario)

It works, but im not sure about it. The branch its in is based on this exclusion branch as well, so it provides

-addon regex exclusion
addon exact match (wont do any further regex matching)
addon not an exact match (then is done using regex matching)

some examples can be seen https://github.com/xbmc/xbmc/issues/17773#issuecomment-620359946

it seems like a hack, as i still have to strip * from anything provided to ensure the filter regex doesnt then match the same as its currently done. Just might be a slight input from things going wild.

The only other way (cleanly) i can come up with would be to change so regex arguments are provides as eg +addonregex, to mirror the -addon for the exclusion state in this rule. Anything without a + or - would then be only considered for exact match. Downside to this, it changes current expected behaviour. At least it would then be explicit in what action a user is wanting however.

fuzzard committed:
[cmake] addons_to_build better regex/exact match handling

Handles regex matching and exact addon matching as separate lists. This allows to
only do regex matches when the exact match has not been found, or an actual regex expression
was provided

Implements an exclusion match regex to exclude an addon/s from building

example usage
ADDONS_TO_BUILD='game.libretro.* -game.libretro.mrboom peripheral.joystick'

game.libretro.mrboom will not build, but game.libretro and game.libretro.2048 will

fuzzard commented:
Cleaned it up slightly and made it compatible with cmake 3.5. If/when we do change minimum cmake version to 3.6+, list filters will be handy to change a lot of our looping over to i think.

https://github.com/fuzzard/xbmc/commits/cmake3.5_addonregexmatching

fuzzard commented:
jenkins build this please

fuzzard commented:
Have added a commit to resolve https://github.com/xbmc/xbmc/issues/17773

Figure best to try to fix everything in the one PR for this.

phunkyfish commented:
Are you happy with this on @AlwinEsch ?

AlwinEsch commented:
Not really, as the most basic no more works `cmake .. -DADDONS_TO_BUILD="screensaver.*"`.
Also seen a bugs inside, give me a bit to make big checks.

fuzzard commented:
Have you tried it since the new commit? Your use case definitely works with the current PR commits. ie https://github.com/xbmc/xbmc/pull/17755/commits/b2a84a842f1c28db0d714aa7d8a41fd9b0a68641

fuzzard commented:
@AlwinEsch have updated the commit to your suggested changes. One minor change that just sets ADDONS_TO_BUILD to all if its not set. Save having to check the empty var and string "all" several times

fuzzard commented:
Jenkins build this please
